File: Cliente2/View/Predio/GUIActualizar.py
import tkinter as tk
from tkinter import messagebox
import requests
import json
from utils import centrar_ventana


def GUIActualizar(root):
    root.title("Actualizar Predio")
    centrar_ventana(root)

    labels = ["ID", "Propietario", "Direccion", "Estrato", "Consumo m3", "Fecha Registro", "Valor factura"]
    entries = {}

    for i, label in enumerate(labels):
        tk.Label(root, text=label).grid(row=i, column=0, padx=10, pady=5, sticky="e")
        entry = tk.Entry(root)
        entry.grid(row=i, column=1, padx=10, pady=5)
        entries[label] = entry

    entries["Fecha Registro"].config(state="readonly")

    def consultar():
        id_ = entries["ID"].get().strip()
        if not id_.isdigit():
            messagebox.showerror("Error", "El ID debe ser un número entero.")
            return

        try:
            response = requests.get(f"http://localhost:8091/predios/{id_}")
            if response.status_code == 200:
                data = response.json()
                entries["Propietario"].delete(0, tk.END)
                entries["Propietario"].insert(0, data["representanteLegal"])
                entries["Direccion"].delete(0, tk.END)
                entries["Direccion"].insert(0, data["direccion"])
                entries["Estrato"].delete(0, tk.END)
                entries["Estrato"].insert(0, data["estrato"])
                entries["Consumo m3"].delete(0, tk.END)
                entries["Consumo m3"].insert(0, data["consumo"])
                entries["Fecha Registro"].config(state="normal")
                entries["Fecha Registro"].delete(0, tk.END)
                entries["Fecha Registro"].insert(0, data["fechaRegistro"])
                entries["Fecha Registro"].config(state="readonly")
                entries["Valor factura"].delete(0, tk.END)
                entries["Valor factura"].insert(0, data["valorFactura"])
            else:
                messagebox.showerror("Error", f"Predio no encontrado. Código: {response.status_code}")
        except Exception as e:
            messagebox.showerror("Error", f"Error al consultar: {e}")

    def actualizar():
        try:
            id_ = entries["ID"].get().strip()
            if not id_.isdigit():
                messagebox.showerror("Error", "El ID debe ser un número entero.")
                return

            estrato = entries["Estrato"].get().strip()
            if not estrato.isdigit() or not (1 <= int(estrato) <= 6):
                messagebox.showerror("Error", "Estrato debe ser un número entre 1 y 6.")
                return

            consumo = entries["Consumo m3"].get().strip()
            if not consumo.replace('.', '', 1).isdigit():
                messagebox.showerror("Error", "Consumo debe ser un número positivo.")
                return

            datos = {
                "codigo": int(id_),
                "representanteLegal": entries["Propietario"].get().strip(),
                "direccion": entries["Direccion"].get().strip(),
                "estrato": int(estrato),
                "consumo": float(consumo)
            }

            response = requests.put(f"http://localhost:8091/predios/{id_}", json=datos)
            if response.status_code == 200:
                messagebox.showinfo("Éxito", "Predio actualizado exitosamente.")
            else:
                messagebox.showerror("Error", f"No se pudo actualizar el predio. Código: {response.status_code}")
        except Exception as e:
            messagebox.showerror("Error", f"Error al actualizar: {e}")

    frame_botones = tk.Frame(root)
    frame_botones.grid(row=len(labels), column=0, columnspan=2, pady=20)
    tk.Button(frame_botones, text="Consultar", command=consultar).pack(side="left", padx=10)
    tk.Button(frame_botones, text="Actualizar", command=actualizar).pack(side="right", padx=10)

File: Cliente/Cliente/GUIActualizar.cs
﻿using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Net.Http;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;
using Newtonsoft.Json;
using RestSharp;

namespace Cliente
{
    public partial class GUIActualizar : Form
    {
        public GUIActualizar()
        {
            InitializeComponent();
            this.StartPosition = FormStartPosition.CenterScreen; // Centrar ventana en la pantalla
            txtFecha.ReadOnly = true;
        }
        private void btnActualizar_Click(object sender, EventArgs e)
        {
            try
            {
                // Validación previa de campos obligatorios
                if (string.IsNullOrWhiteSpace(txtCodigo.Text) ||
                    string.IsNullOrWhiteSpace(txtRepresentanteLegal.Text) ||
                    string.IsNullOrWhiteSpace(txtDireccion.Text) ||
                    //string.IsNullOrWhiteSpace(txtEstado.Text) ||
                    string.IsNullOrWhiteSpace(txtEstrato.Text) ||
                    string.IsNullOrWhiteSpace(txtConsumo.Text) ||
                    string.IsNullOrWhiteSpace(txtValorFactura.Text))
                    //string.IsNullOrWhiteSpace(txtSubsidio.Text) ||
                    //string.IsNullOrWhiteSpace(txtComercio.Text))
                {
                    MessageBox.Show("Por favor complete todos los campos antes de continuar.", "Campos incompletos", MessageBoxButtons.OK, MessageBoxIcon.Warning);
                    return;
                }

                // Validar ID
                if (!int.TryParse(txtCodigo.Text.Trim(), out int id))
                {
                    MessageBox.Show("El ID debe ser un número entero.", "Error de formato", MessageBoxButtons.OK, MessageBoxIcon.Error);
                    return;
                }

                // Validar Propietario (solo letras)
                string representanteLegal = txtRepresentanteLegal.Text.Trim();
                if (!representanteLegal.All(char.IsLetter))
                {
                    MessageBox.Show("El campo 'Propietario' solo debe contener letras.", "Error de formato", MessageBoxButtons.OK, MessageBoxIcon.Error);
                    return;
                }

                // Validar EstadoCuenta (AC o INAC)
                /*string estado = txtEstado.Text.Trim().ToUpper();
                if (estado != "AC" && estado != "INAC")
                {
                    MessageBox.Show("El campo 'Estado de Cuenta' debe ser 'AC' o 'INAC'.", "Error de formato", MessageBoxButtons.OK, MessageBoxIcon.Error);
                    return;
                }*/

                // Validar y convertir Estrato
                if (!int.TryParse(txtEstrato.Text.Trim(), out int estrato))
                {
                    MessageBox.Show("El campo 'Estrato' debe ser un número entero válido.", "Error de formato", MessageBoxButtons.OK, MessageBoxIcon.Error);
                    return;
                }

                if (estrato < 1 || estrato > 6)
                {
                    MessageBox.Show("El campo 'Estrato' debe estar entre 1 y 6.", "Error de valor", MessageBoxButtons.OK, MessageBoxIcon.Error);
                    return;
                }

                // Validar Consumo
                if (!double.TryParse(txtConsumo.Text.Trim(), out double consumo) || consumo < 0)
                {
                    MessageBox.Show("El campo 'Consumo' debe ser un número positivo.", "Error de formato", MessageBoxButtons.OK, MessageBoxIcon.Error);
                    return;
                }

                // Validar Subsidio
                /*if (!int.TryParse(txtSubsidio.Text.Trim(), out int subsidio) || subsidio < 0)
                {
                    MessageBox.Show("El campo 'Subsidio' debe ser un número entero positivo.", "Error de formato", MessageBoxButtons.OK, MessageBoxIcon.Error);
                    return;
                }*/

                // Validar TipoVivienda (string no vacío)
                /*string tipoComercio = txtComercio.Text.Trim();
                if (string.IsNullOrWhiteSpace(tipoComercio))
                {
                    MessageBox.Show("El campo 'Tipo de Vivienda' no puede estar vacío.", "Campos incompletos", MessageBoxButtons.OK, MessageBoxIcon.Warning);
                    return;
                }*/

                string direccion = txtDireccion.Text.Trim();

                // Crear el objeto Request
                var requestData = new Request
                {
                    codigo = id,
                    representanteLegal = representanteLegal,
                    direccion = direccion,
                    //EstadoCuenta = estado,
                    estrato = estrato,
                    consumo = consumo,
                    //Subsidio = subsidio,
                    //TipoComercio = tipoComercio
                };

                // Crear cliente REST
                var options = new RestClientOptions("http://localhost:8091");
                var client = new RestClient(options);

                // Crear solicitud PUT
                var request = new RestRequest($"/predios/{id}", Method.Put);
                request.AddJsonBody(requestData);

                // Enviar solicitud
                var response = client.Execute(request);

                // Verificar respuesta
                if (response.StatusCode == System.Net.HttpStatusCode.OK)
                {
                    MessageBox.Show("Predio actualizado exitosamente.", "Éxito", MessageBoxButtons.OK, MessageBoxIcon.Information);
                }
                else
                {
                    MessageBox.Show("Error al actualizar predio. Código: " + (int)response.StatusCode + "\nDetalle: " + response.Content, "Error en respuesta", MessageBoxButtons.OK, MessageBoxIcon.Error);
                }
            }
            catch (HttpRequestException ex)
            {
                MessageBox.Show("Error de conexión al servidor: " + ex.Message, "Error de red", MessageBoxButtons.OK, MessageBoxIcon.Error);
            }
            catch (FormatException ex)
            {
                MessageBox.Show("Error de formato: " + ex.Message, "Error de datos", MessageBoxButtons.OK, MessageBoxIcon.Error);
            }
            catch (Exception ex)
            {
                MessageBox.Show("Error inesperado: " + ex.Message, "Error general", MessageBoxButtons.OK, MessageBoxIcon.Error);
            }
        }


        private void btnConsultar_Click_1(object sender, EventArgs e)
        {


            try
            {
                if (!int.TryParse(txtCodigo.Text.Trim(), out int id))
                {
                    MessageBox.Show("Error: El ID debe ser un número entero.");
                    return;
                }

                var options = new RestClientOptions("http://localhost:8091");
                var client = new RestClient(options);

                var request = new RestRequest($"/predios/{id}", Method.Get);

                var response = client.Execute(request);

                if (response.StatusCode == System.Net.HttpStatusCode.OK)
                {
                    var predio = JsonConvert.DeserializeObject<Residencial>(response.Content);
                    txtRepresentanteLegal.Text = predio.representanteLegal;
                    txtDireccion.Text = predio.direccion;
                    txtFecha.Text = predio.fechaRegistro.ToString("yyyy-MM-ddTHH:mm:ss");
                    //txtEstado.Text = predio.EstadoCuenta;
                    txtEstrato.Text = predio.estrato.ToString();
                    txtConsumo.Text = predio.consumo.ToString();
                    txtValorFactura.Text = predio.valorFactura.ToString();
                    //txtSubsidio.Text = predio.Subsidio.ToString();
                    //txtComercio.Text = predio.TipoComercio;
                }
                else
                {
                    MessageBox.Show("Error al consultar predio: " + response.StatusCode + "\nContenido: " + response.Content);
                }
            }
            catch (Exception ex)
            {
                MessageBox.Show("Error inesperado: " + ex.Message);
            }

        }



        private void txtDireccion_TextChanged(object sender, EventArgs e)
        {

        }

        private void txtConsumo_TextChanged(object sender, EventArgs e)
        {

        }
    }
}